refactor: update and turn `WsTransportClientBuilder` generic

oleonardolima opened the pull request:
e7ecb9d5bc41ee0cb018a0d4a97c20ed8cb46966 refactor(jsonrpsee-client-transport): update and turn `WsTransportClientBuilder` generic 

- fix(docs): typo on `WsTransportClientBuilder` doc on `Receiver`
  reference.
- refactor: add initial fn signatures for `build_with_stream`,
  `try_connect_over_tcp`, and `try_connect`.
- refactor: expose `EitherStream` visibility to public.
- refactor: make `Sender` and `Receiver` generic over T, a data stream.
- refactor: make `TransportSenderT` and `TransportReceiverT`
  implementations over generic `Sender` and `Receiver`, bound to
`AsyncRead`, `AsyncRead`, `MaybeSend` and `'static`.
- refactor: turn old `try_connect` TCP steps into
  `try_connect_over_tcp`.
- feat: implement `build_with_stream` and `try_connect` to handle and
  handle the handshake for a generic data stream `T`.
- feat: add new `Redirected` error variant to `WsHandshakeError`, as it
  should be handled by the client when using a generic data stream `T`.

c7ae7dd79c781dc87e169838c0336c06b1f5e62f refactor(jsonrpsee-ws-client): add new fns to WsClientBuilder 

- feat: add new `WsClientBuilder::build_with_transport` that builds and
  returns a `WsClient` with the given `Sender` and `Receiver`.
- feat: add new `WsClientBuilder::build_with_stream` that uses the new
  `WsTransportClientBuilder::build_with_stream`, building and returning
  the `WsClient` with the given `data_stream` as transport layer.
- refactor: update the `WsClientBuilder::build` to use the new
  `build_with_transport`, it helps not having duplicated code.
  
10e78a045cc07d5b39f8609f13a7ace8284e741f refactor: re-export EitherStream & sort current list

53641d1ec73c0f4a17191cd7f32b52f646226895 test: add integration tests and helper fns

- add new helper fns to spawn a socks5 server, using `fast-socks5`
- add new helper enum for `DataStream` that acts as a wrapper to
  `Socks5Stream<T>`, similar to what a client would need to do.
- impl AsyncRead + AsyncWrite for the helper `DataStream` enum, to make
  it compatible between futures::io and tokio::io.
- add new tests that connects over a socks5 proxy, and use the new
  `WsClientBuilder::default()::build_with_stream(...) fn.


fixes #1162 #870

oleonardolima committed:
refactor(jsonrpsee-client-transport): update and turn `WsTransportClientBuilder` generic

- fix(docs): typo on `WsTransportClientBuilder` doc on `Receiver`
  reference.
- refactor: add initial fn signatures for `build_with_stream`,
  `try_connect_over_tcp`, and `try_connect`.
- refactor: expose `EitherStream` visibility to public.
- refactor: make `Sender` and `Receiver` generic over T, a data stream.
- refactor: make `TransportSenderT` and `TransportReceiverT`
  implementations over generic `Sender` and `Receiver`, bound to
`AsyncRead`, `AsyncRead`, `MaybeSend` and `'static`.
- refactor: turn old `try_connect` TCP steps into
  `try_connect_over_tcp`.
- feat: implement `build_with_stream` and `try_connect` to handle and
  handle the handshake for a generic data stream `T`.
- feat: add new `Redirected` error variant to `WsHandshakeError`, as it
  should be handled by the client when using a generic data stream `T`.
- TODO(@oleonardolima): Add new tests that uses a different data stream.

niklasad1 commented:
@oleonardolima any update on this one? 

Let us know whether you need help to push this further.

oleonardolima commented:
@niklasad1 I plan to get back to it by the end of the week. Sorry for the delayed PR, conferences and changing jobs got me delayed on it.

oleonardolima committed:
refactor(jsonrpsee-ws-client): add new fns to `WsClientBuilder`

- feat: add new `WsClientBuilder::build_with_transport` that builds and
  returns a `WsClient` with the given `Sender` and `Receiver`.
- feat: add new `WsClientBuilder::build_with_stream` that uses the new
  `WsTransportClientBuilder::build_with_stream`, building and returning
  the `WsClient` with the given `data_stream` as transport layer.
- refactor: update the `WsClientBuilder::build` to use the new
  `build_with_transport`, it helps not having duplicated code.

oleonardolima commented:
> @oleonardolima any update on this one?
> 
> Let us know whether you need help to push this further.

@niklasad1 Do you see any problem with adding the [arti-client](https://crates.io/crates/arti-client) dependency on `jsonrpsee-integration-tests` in order to have integration tests for streams other than `tokio::net::TcpStream` ?

niklasad1 commented on client/transport/src/ws/mod.rs:
```diff
@@ -404,6 +400,67 @@ impl WsTransportClientBuilder {
 	}
+
+	/// Try to establish the connection over the given data stream.
+	pub async fn build_with_stream<T>(
```
This should be added to the `WsClientBuilderBuilder` for a uniform API I reckon.

niklasad1 replied:
ah my bad it was added on that type

niklasad1 commented on proc-macros/tests/ui/incorrect/rpc/rpc_deprecated_method.stderr:
```diff
@@ -1,11 +1,11 @@
    |
 note: the lint level is defined here
-  --> $DIR/rpc_deprecated_method.rs:4:9
+  --> tests/ui/incorrect/rpc/rpc_deprecated_method.rs:4:9
```
please revert the absolute paths here

```suggestion
  --> $DIR/rpc_deprecated_method.rs:4:9
```

niklasad1 commented on proc-macros/tests/ui/incorrect/rpc/rpc_deprecated_method.stderr:
```diff
@@ -1,11 +1,11 @@
-error: use of deprecated method `DeprecatedClient::async_method`: please use `new_method` instead
-  --> $DIR/rpc_deprecated_method.rs:63:20
+error: use of deprecated associated function `DeprecatedClient::async_method`: please use `new_method` instead
+  --> tests/ui/incorrect/rpc/rpc_deprecated_method.rs:63:20
```
```suggestion
  --> $DIR/rpc_deprecated_method.rs:63:20
```

niklasad1 commented on proc-macros/tests/ui/incorrect/rpc/rpc_deprecated_method.stderr:
```diff
@@ -1,11 +1,11 @@
-error: use of deprecated method `DeprecatedClient::async_method`: please use `new_method` instead
-  --> $DIR/rpc_deprecated_method.rs:63:20
+error: use of deprecated associated function `DeprecatedClient::async_method`: please use `new_method` instead
```
Which rustc version are you using to get these? 

This repo is using rustc stable and I guess you are using nightly or something?

oleonardolima replied:
I was getting some errors during the build because of the rustc version, and I wrongly used the TRYBUILD=overwrite flag 😅, now it's fixed.

niklasad1 commented:
> @niklasad1 Do you see any problem with adding the [arti-client](https://crates.io/crates/arti-client) dependency on jsonrpsee-integration-tests in order to have integration tests for different streams other than tokio::net::TcpStream ?

It's not possible to use a more light-weight dependency than arti-client for that? 

I'm happy if we just have a test to wrap another stream using the WsTransportBuilder and assert connecting and make a simple RPC call works.

niklasad1 commented on client/transport/src/ws/mod.rs:
```diff
@@ -276,12 +292,18 @@ impl TransportReceiverT for Receiver {
-		self.try_connect(target).await
+	///
+	/// Uses the default connection over TCP.
+	pub async fn build(self, uri: Url) -> Result<(Sender<EitherStream>, Receiver<EitherStream>), WsHandshakeError> {
```
please add `build_with_stream` here

niklasad1 commented on client/ws-client/src/lib.rs:
```diff
@@ -260,4 +248,55 @@ impl WsClientBuilder {
+		Ok(ws_client)
+	}
+
+	/// Build the [`WsClient`] with specified URL to connect t, using the default
```
```suggestion
	/// Build the [`WsClient`] with specified URL to connect to, using the default
```

niklasad1 commented on client/ws-client/src/lib.rs:
```diff
@@ -260,4 +248,55 @@ impl WsClientBuilder {
 	}
+
+	/// Build the [`WsClient`] with specified data stream, using [`WsTransportClientBuilder::build_with_stream`].
+	/// You must provide the port number in the URL.
```
this have been fixed recently, so remove that comment

```suggestion
```

niklasad1 commented on client/ws-client/src/lib.rs:
```diff
@@ -260,4 +248,55 @@ impl WsClientBuilder {
+
+	/// Build the [`WsClient`] with specified URL to connect t, using the default
+	/// [`WsTransportClientBuilder::build_with_stream`], therefore with the default TCP as transport layer.
+	/// You must provide the port number in the URL.
```
```suggestion
```

niklasad1 reviewed (commented):
Yupp, nice looks you are on the right path.

This PR just needs to fix a few minor things + add few tests then it should be good to go.

oleonardolima commented:
> It's not possible to use a more light-weight dependency than arti-client for that?
> 
> I'm happy if we just have a test to wrap another stream using the `WsTransportBuilder::build_with_stream` and assert connecting and make a simple RPC call works.

Sure, I'll check if there is a more light-weight one.

_PS.: We can use the TcpStream, but wouldn't be testing anything new in my opinion (because is the default already used one)_

niklasad1 commented:
> PS.: We can use the TcpStream, but wouldn't be testing anything new in my opinion (because is the default already used one)

Yeah, I agree that we should test with another one

oleonardolima committed:
refactor: re-export `EitherStream` & sort current list

oleonardolima committed:
test: add integration tests and helper fns

- add new helper fns to spawn a socks5 server, using `fast-socks5`
- add new helper enum for `DataStream` that acts as a wrapper to
  `Socks5Stream<T>`, similar to what a client would need to do.
- impl AsyncRead + AsyncWrite for the helper `DataStream` enum, to make
  it compatible between futures::io and tokio::io.
- add new tests that connects over a socks5 proxy, and use the new
  `WsClientBuilder::default()::build_with_stream(...) fn.

niklasad1 commented on tests/tests/helpers.rs:
```diff
@@ -249,3 +253,89 @@ pub async fn pipe_from_stream_and_drop<T: Serialize>(
+
+#[pin_project(project = DataStreamProj)]
+#[allow(dead_code)]
+pub enum DataStream<T: tokio::io::AsyncRead + tokio::io::AsyncWrite + std::marker::Unpin> {
```
I don't understand why this type is needed, can't you just use `fast_socks5::client::Socks5Stream` directly instead?

It already implements AsyncRead and AsyncWrite?

oleonardolima replied:
@niklasad1 It does implement the `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`, but currently the [`TransportSenderT` and `TransportReceiverT`](https://github.com/paritytech/jsonrpsee/pull/1168/files#diff-a8985d79c808da60a2c0bd792d8d0e303791e5c1618215e764f582d8c75de292R228-R271) requires the `futures::io::AsyncRead` and `futures::io::AsyncRead` implementation.

I made this way not to change the current implementation of [`EitherStream`](https://github.com/paritytech/jsonrpsee/blob/ac0caafd4edae89f4d0aab15ccfedb93cb927dfc/client/transport/src/ws/stream.rs#L41-L154), which implements the `futures::io` traits. 

I can change the `EitherStream` to use the `tokio::io` traits instead, do you see any problem ?

> I'm actually biased because requiring the tokio traits as default would make it easier for me on other projects 😅

niklasad1 replied:
This is just for tests so I don't mind having this extra code.

> I can change the EitherStream to use the tokio::io traits instead, do you see any problem ?

Let's keep it out this PR but `tokio::io` makes sense to me since this crate it's tightly-coupled to tokio nowdays.
This is probably old code when this was supposed to support both tokio and async-std.

//cc @lexnv @jsdw

jsdw replied:
Agreed that moving to `tokio::io` traits probably makes sense and might simplify some stuff given that we are "all in" on tokio in this lib nowadays

jsdw replied:
Also I wonder why this is an enum and not struct?

niklasad1 replied:
yeah, would be cleaner with a struct to avoid matching on it :P

oleonardolima replied:
Great, thanks for the suggestion and @niklasad1 for addressing the changes

niklasad1 reviewed (commented):
Nice, looks good

Run `cargo fmt --all` to make the CI happy.

niklasad1 reviewed (approved):
Looks good to me, nice work

Can you please run `cargo clippy` and fix the warnings?

oleonardolima commented:
> Looks good to me, nice work
> 
> Can you please run `cargo clippy` and fix the warnings?

Sure, CI is all good now 🚀

jsdw commented on client/transport/src/ws/mod.rs:
```diff
@@ -396,14 +391,71 @@ impl WsTransportClientBuilder {
 	}
+
+	/// Try to establish the connection over the given data stream.
+	pub async fn build_with_stream<T>(
```
Nit: I'd move this fn to be right underneath `build`

jsdw commented on tests/tests/helpers.rs:
```diff
@@ -249,3 +254,112 @@ pub async fn pipe_from_stream_and_drop<T: Serialize>(
 	}
 }
+
+#[allow(dead_code)]
```
I'd have to compile this to check but I wonder why all the `#[allow(dead_code)]`s are needed!

niklasad1 replied:
I don't remember but probably that is just used for tests are not exported outside the crate.

Maybe pub(crate) would do it...

jsdw replied:
If they are only used for tests then having the helpers be behind `#[cfg(test)]` would also remove such a warning :)

niklasad1 replied:
we don't have a lib.rs in the crate it's like https://github.com/paritytech/jsonrpsee/tree/master/tests/tests but every file should #![cfg(test)] so I don't know ^^

oleonardolima replied:
Sorry about that, totally missed it

jsdw commented on tests/tests/helpers.rs:
```diff
@@ -249,3 +254,112 @@ pub async fn pipe_from_stream_and_drop<T: Serialize>(
+	}
+}
+
+impl<T: tokio::io::AsyncRead + tokio::io::AsyncWrite + std::marker::Unpin> AsyncWrite for DataStream<T> {
```
Nit: `T` doesn't need AsyncRead?

niklasad1 replied:
same here

jsdw commented on tests/tests/helpers.rs:
```diff
@@ -249,3 +254,112 @@ pub async fn pipe_from_stream_and_drop<T: Serialize>(
+	Socks5(#[pin] Socks5Stream<T>),
+}
+
+impl<T: tokio::io::AsyncRead + tokio::io::AsyncWrite + std::marker::Unpin> AsyncRead for DataStream<T> {
```
Nit: `T` doesn't need AsyncWrite bound?

niklasad1 replied:
nah, it doesn't work the Socks5Stream has trait bound on the type itself to T: AsyncRead + AsyncWrite + Unpin.

Thus, these AsyncWrite can't be omitted even if it shouldn't be needed

jsdw commented on tests/tests/integration_tests.rs:
```diff
@@ -108,6 +133,47 @@ async fn ws_unsubscription_works() {
+
+	sub.unsubscribe().await.unwrap();
+
+	let mut success = false;
+
+	// Wait until a slot is available, as only one concurrent call is allowed.
+	// Then when this finishes we know that unsubscribe call has been finished.
+	for _ in 0..30 {
+		let res: Result<String, _> = client.request("say_hello", rpc_params![]).await;
+		if res.is_ok() {
+			success = true;
+			break;
+		}
+		tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+	}
```
So this is all just here to check that unsubscribe has finished; I wonder whether there is a simpler/cleaner way to do this (@niklasad1)

niklasad1 replied:
yeah but I realized that the subscription doesn't keep the `request id guard` so this doesn't work as intended.

let's remove extra ugliness for the test...

jsdw reviewed (approved):
Looks like a nice addition to me; thank you! Just some nits but nothing super important

niklasad1 committed:
address review suggestions
